appcelerator/titanium_mobile
# feat(android): Optimized TextField/TextArea handling in TableView

**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-26862

**Summary:**
- Modified `TableView` to no longer recycle row container view.
  * There is no point in doing this since we always keep all child views in memory. Very little memory savings. (Use `ListView` for this behavior instead.)
  * Changing the parent view for rows containing `TextField` and `TextArea` was causing havoc while the virtual keyboard was shown.
- There is no good way to remove the keyboard warnings caused by changing a `TextField` or `TextArea` value/text programmatically. We have to live with it, but I've confirmed that it doesn't cause performance issues either.

---
**Test TableView with TextAreas/TextFields:**
_This tests the optimization change made._
1. Build and run code `TableOfTextFieldsTest.js` attached to [TIMOB-26862](https://jira.appcelerator.org/browse/TIMOB-26862) on Android 7.0 or higher.
2. Note that the first 10 rows contains `TextAreas` and the remaining rows contain `TextFields`.
3. Scroll to the top of the `TableView`.
4. Tap on row 1's `TextArea`.
5. Verify that all of the text within the `TextArea` is highlight.
6. Type in "abc".
7. Verify that the text is changed to "ABC" as you type it in.
8. Dismiss the virtual keyboard. (ie: Press the "Back" button once.)
9. Scroll to the bottom of the `TableView`.
10. Tap on row 200's `TextField`.
11. Verify that all of the text with the `TextField` is highlighted.
12. Type in "xyz".
13. Verify that the text is changed to "XYZ" as you type it in.
14. Dismiss the virtual keyboard.
15. Scroll to the top of the `TableView`.
16. Verify that row 1 still reads "ABC".
17. Repeat steps 1-16 on Android 4.4.

---
**Test TableView Default Rows:**
_This tests that "default" row template updates/handling still works._
1. Build and run code `TableViewDefaultRowTest.js` attached to [TIMOB-26862](https://jira.appcelerator.org/browse/TIMOB-26862) on Android 7.0 or higher.
2. Verify that row background colors change every 2 second to red, orange, and blue.
3. Verify that each row increments its "Count X" text updates every 2 seconds.
4. Verify that every row's left and right icons cycle between the appc logo to a soccer ball every 2 seconds.
5. Scroll to the bottom of the `TableView`.
6. Verify steps 2-4 above for the bottom rows.
7. Repeat steps 1-6 on Android 4.4.

---
**Test TableView Custom Rows:**
_This tests that row child view changes still works._
1. Build and run code `TableViewCustomRowTest.js` attached to [TIMOB-26862](https://jira.appcelerator.org/browse/TIMOB-26862) on Android 7.0 or higher.
2. Verify that row background colors change every 2 second to red, orange, and blue.
3. Verify that each row increments its "Count X" text updates every 2 seconds.
4. Verify that every row's left and right icons cycle between the appc logo to a soccer ball every 2 seconds.
5. Scroll to the bottom of the `TableView`.
6. Verify steps 2-4 above for the bottom rows.
7. Repeat steps 1-6 on Android 4.4.
